dbms/src/Storages/DeltaMerge/tests/gtest_local_indexer_scheduler.cpp

```diff
@@ -0,0 +1,541 @@
+
+    scheduler->start();
+    scheduler->waitForFinish();
+
+    // Scheduler is scheduled by KeyspaceID asc order and TableID asc order.
+    ASSERT_EQ(results.size(), 5);
+    ASSERT_STREQ(results[0].c_str(), "ks1_t1_#1");
+    ASSERT_STREQ(results[1].c_str(), "ks2_t1_#1");
+    ASSERT_STREQ(results[2].c_str(), "ks1_t2_#1");
+    ASSERT_STREQ(results[3].c_str(), "ks1_t3_#1");
+    ASSERT_STREQ(results[4].c_str(), "ks1_t3_#2");
+}
+CATCH
+
+TEST_F(LocalIndexerSchedulerTest, TaskExceedMemoryLimit)
+try
+{
+    auto scheduler = LocalIndexerScheduler::create({
+        .pool_size = 10,
+        .memory_limit = 2,
+        .auto_start = false,
+    });
+
+    ASSERT_THROW(
+        {
+            scheduler->pushTask({
+                .keyspace_id = 1,
+                .table_id = 1,
+                .file_ids = {LocalIndexerScheduler::DMFileID(1)},
+                .request_memory = 100,
+                .workload = [&]() { pushResult("foo"); },
+            });
+        },
+        DB::Exception);
+    ASSERT_NO_THROW({
+        scheduler->pushTask({
+            .keyspace_id = 1,
+            .table_id = 1,
+            .file_ids = {LocalIndexerScheduler::DMFileID(2)},
+            .request_memory = 0,
+            .workload = [&]() { pushResult("bar"); },
+        });
+    });
+
+    scheduler->start();
+    scheduler->waitForFinish();
+
+    ASSERT_EQ(results.size(), 1);
+    ASSERT_STREQ(results[0].c_str(), "bar");
+
+    results.clear();
+
+    scheduler = LocalIndexerScheduler::create({
+        .pool_size = 10,
+        .memory_limit = 0,
+    });
+
+    ASSERT_NO_THROW({
+        scheduler->pushTask({
+            .keyspace_id = 1,
+            .table_id = 1,
+            .file_ids = {LocalIndexerScheduler::DMFileID(3)},
+            .request_memory = 100,
+            .workload = [&]() { pushResult("foo"); },
+        });
+    });
+    ASSERT_NO_THROW({
+        scheduler->pushTask({
+            .keyspace_id = 1,
+            .table_id = 1,
+            .file_ids = {LocalIndexerScheduler::DMFileID(4)},
+            .request_memory = 0,
+            .workload = [&]() { pushResult("bar"); },
+        });
+    });
+
+    scheduler->start();
+    scheduler->waitForFinish();
+
+    ASSERT_EQ(results.size(), 2);
+    ASSERT_STREQ(results[0].c_str(), "foo");
+    ASSERT_STREQ(results[1].c_str(), "bar");
+}
+CATCH
+
+TEST_F(LocalIndexerSchedulerTest, MemoryLimit)
+try
+{
+    auto scheduler = LocalIndexerScheduler::create({
+        .pool_size = 10,
+        .memory_limit = 2,
+        .auto_start = false,
+    });
+
+    auto task_1_is_started = std::make_shared<std::promise<void>>();
+    auto task_2_is_started = std::make_shared<std::promise<void>>();
+    auto task_3_is_started = std::make_shared<std::promise<void>>();
+
+    auto task_1_wait = std::make_shared<std::promise<void>>();
+    auto task_2_wait = std::make_shared<std::promise<void>>();
+    auto task_3_wait = std::make_shared<std::promise<void>>();
+
+    scheduler->pushTask({
```

Comment:
I wonder if `fid=1` is not in `adding_index_page_id_set`, how can we tell if it is not added, or if it is already ready?
For example, if thre is only one task `(ks=1, tid=1, fid=1)`. Could we `addTask(ks=1, tid=1, fid=1)` and then `addTask(ks=1, tid=1, fid=1)` again?

Reply:
Yes, we support adding the same task multiple times, but they are not allowed to execute at the same time.

Reply:
Is there any tests to explain this behavior? I feel it hard to understand

Reply:
```cpp
TEST_F(LocalIndexerSchedulerTest, FileIsUsing)
```

Reply:
> Yes, we support adding the same task multiple times, but they are not allowed to execute at the same time.

@Lloyd-Pottiger Pls add comments about this to the `pushTask`